keras_nlp/models/llama/llama_attention.py

```diff
@@ -0,0 +1,197 @@
+
+        self._kernel_initializer = keras.initializers.get(kernel_initializer)
+        self._bias_initializer = keras.initializers.get(bias_initializer)
+        self.max_sequence_length = max_sequence_length
+
+        self.rope_scaling_type = rope_scaling_type
+        self.rope_scaling_factor = rope_scaling_factor
+
+        if rope_scaling_type is None:
+            self.rotary_embedding_layer = RotaryEmbedding(
+                max_wavelength=rotary_max_wavelength
+            )
+        else:
+            if rope_scaling_type == "linear":
+                self.rotary_embedding_layer = RotaryEmbedding(
+                    max_wavelength=rotary_max_wavelength,
+                    scaling_factor=rope_scaling_factor,
+                )
+            else:
+                raise ValueError(
+                    f"Unknown RoPE scaling type {rope_scaling_type}"
+                )
+
+    def build(self, inputs_shape):
+        self.hidden_dim = inputs_shape[-1]
+        self.attn_head_size = self.hidden_dim // self.num_query_heads
+
+        self._query_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_query_heads, self.attn_head_size),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="query",
+        )
+        self._query_dense.build(inputs_shape)
+
+        self._key_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads, self.attn_head_size),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="key",
+        )
+        self._key_dense.build(inputs_shape)
+
+        self._value_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads, self.attn_head_size),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="value",
+        )
+        self._value_dense.build(inputs_shape)
+
+        self._softmax = keras.layers.Softmax(axis=-1, name="attention_softmax")
+
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="attention_output",
+        )
+        self._output_dense.build(inputs_shape)
+
+        self.built = True
+
+    def call(
+        self, hidden_states, attention_mask=None, return_attention_scores=False
+    ):
+        query = self._query_dense(hidden_states)
+        key = self._key_dense(hidden_states)
+        value = self._value_dense(hidden_states)
+
+        query = self.rotary_embedding_layer(query)
+        key = self.rotary_embedding_layer(key)
+
+        key = ops.tile(key, [1, 1, self.num_key_value_groups, 1])
+        value = ops.tile(value, [1, 1, self.num_key_value_groups, 1])
```

Comment:
The original model used `torch.repeat_interleave` which I think does a different operatrion than `ops.tile`:

```python
from keras import ops

ops.repeat([1,2,3], 2)
# <tf.Tensor: shape=(6,), dtype=int32, numpy=array([1, 1, 2, 2, 3, 3], dtype=int32)>

ops.tile([1,2,3], [2])
# <tf.Tensor: shape=(6,), dtype=int32, numpy=array([1, 2, 3, 1, 2, 3], dtype=int32)>

import torch

torch.repeat_interleave(torch.tensor([1,2,3]), 2)
# tensor([1, 1, 2, 2, 3, 3])
```

Any reason to use `ops.tile` over `ops.repeat`?